feat: catching any input errors in the frontend

Dominik1999 opened the pull request:
Closes #18

Dominik1999 commented:
Hmm added a new formatter. Only changes to `App.tsx` are relevant for that PR. The rest is formatting. 

I can change the PR only to contain the changes and then send a PR with the new formatting.

grjte commented:
> Hmm added a new formatter. Only changes to `App.tsx` are relevant for that PR. The rest is formatting.
> 
> I can change the PR only to contain the changes and then send a PR with the new formatting.

That would be great, thanks. It'll be easier to review & a cleaner git history to keep them separate

grjte commented on playground/src/App.tsx:
```diff
@@ -96,26 +174,34 @@ Cycles = <available with Miden VM v0.4>`);
-      init().then(() => {
-        try {
-          const resp = prove_program(code, inputs, numOfOutputs);
-          console.log(resp.toString());
-          const stack_output = resp.slice(0, numOfOutputs);
-          const overflow_addrs = resp.slice(numOfOutputs, resp.length);
-          setOutput(`Miden VM Program Output
+  const proveProgram = async () => {
+    init().then(() => {
+      try {
+        if (!tryParseJSONObject(inputs)) {
+          setOutput(`Error: Invalid JSON input`);
+          return;
+        }
+        if (!checkStackInitAndAdviceTape(inputs)) {
+          setOutput(`Error: Invalid inputs - no stack_init or advice_tape`);
+          return;
+        }
+        const resp = prove_program(code, inputs, numOfOutputs);
+        console.log(resp.toString());
+        const stack_output = resp.slice(0, numOfOutputs);
+        const overflow_addrs = resp.slice(numOfOutputs, resp.length);
+        setOutput(`Miden VM Program Output
 Stack = [${stack_output.toString()}]
 Overflow Address = [${overflow_addrs.toString()}]
 Cycles = <available with Miden VM v0.4>`);
-        } catch (error) {
-          setOutput("Error: Check the developer console for details.");
-        }
-          })};
+      } catch (error) {
+        setOutput("Error: Check the developer console for details.");
+      }
+        })};
     
   /**
   * We need to add the following:
   * Verify, and Debug as functions that can be called.
-  */ 
+  */ ``
```
Why do we need these?

grjte commented on playground/src/App.tsx:
```diff
@@ -96,26 +174,34 @@ Cycles = <available with Miden VM v0.4>`);
+          return;
+        }
+        const resp = prove_program(code, inputs, numOfOutputs);
+        console.log(resp.toString());
```
We can probably remove this console log

Dominik1999 replied:
jup

grjte commented on playground/src/App.tsx:
```diff
@@ -96,26 +174,34 @@ Cycles = <available with Miden VM v0.4>`);
+  const proveProgram = async () => {
+    init().then(() => {
+      try {
+        if (!tryParseJSONObject(inputs)) {
+          setOutput(`Error: Invalid JSON input`);
+          return;
+        }
+        if (!checkStackInitAndAdviceTape(inputs)) {
+          setOutput(`Error: Invalid inputs - no stack_init or advice_tape`);
+          return;
+        }
```
This code is duplicated above. Let's make a helper

Dominik1999 replied:
better?

grjte commented on playground/src/App.tsx:
```diff
@@ -16,6 +17,75 @@ async function getExample(example: string) {
+    return true;
+  }
+  try {
+      const o = JSON.parse(jsonString);
+
+      // Handle non-exception-throwing cases:
+      // Neither JSON.parse(false) or JSON.parse(1234) throw errors, hence the type-checking,
+      // but... JSON.parse(null) returns null, and typeof null === "object", 
+      // so we must check for that, too. Thankfully, null is falsey, so this suffices:
+      if (o && typeof o === "object") {
+          return true;
+      }
+  }
+  catch (e) {
+    toast.error('Miden VM Inputs need to be a valid JSON object.');
+  }
+  
+  return false;
+
+}
+
+/**
+* We allow an empty input.
+* If it is not empty, we check whether the input has 
+* either stack_init or advice_tape.
+* Then we check whether the values are numbers.
+*/ 
+
+function checkStackInitAndAdviceTape(jsonString: string) { 
+
+  if (jsonString === "") { 
+    return true 
+  }
+
+  if (!jsonString.includes("stack_init") && !jsonString.includes("advice_tape")) {
+    toast.error('Miden VM Inputs can be empty or we need either a stack_init or an advice_tape.');
+    return false
+  }
+
+  const o = JSON.parse(jsonString);
```
There's some duplication between these. Is there a reason we can't combine them into a single `checkInputs` function?

Dominik1999 replied:
jup that works as well

grjte reviewed (changes requested):
Thanks @Dominik1999! I didn't try running it up locally yet, so this review is just based on looking at the code. Once you've addressed these comments I can take a quick look at that too

Dominik1999 commented:
> Thanks @Dominik1999! I didn't try running it up locally yet, so this review is just based on looking at the code. Once you've addressed these comments I can take a quick look at that too

Thanks for your review. Your points should be addressed by now. Let me know what you think. 

I tested it locally, but we need to add proper testing to the whole playground, which will be reflected in the next PR - issue here https://github.com/0xPolygonMiden/examples/issues/28

grjte commented on playground/src/App.tsx:
```diff
@@ -16,6 +17,79 @@ async function getExample(example: string) {
+
+  }
+
+  const o = tryParseJSONObject(jsonString);
+
+  if (o === false) {
+    return false;
+  }
```
It looks like this if statement wouldn't be necessary if the contents of `tryParseJSONObject` were moved in here. Given that it's not a long function and it's not used anywhere else (unless I missed it), I would probably prefer to move that completely into here.

I think a few other things could probably be condensed a bit here too. For example, we're doing the checks to see if the string includes "stack_init" and "advice_tape" twice.

Dominik1999 replied:
Got it, refactored quite a bit.

grjte reviewed (changes requested):
Thanks! A few similar additional comments. Also, we shouldn't include the `.DS_Store` file

Overcastan reviewed (approved):
Looks good to me, thank you @Dominik1999! 
Just one small nit

Dominik1999 commented on playground/src/App.tsx:
```diff
@@ -16,6 +17,79 @@ async function getExample(example: string) {
+    return false;
+  }
+
+  if (jsonString.includes("stack_init")) {
+    if (o.stack_init.len === 0 || o.stack_init.some(isNaN)) {
+      
+      toast.error(`If you use stack_init, 
+      it must contain at least one number, 
+      and can only contain numbers.`);
+      
+      return false;
+    }
+  }
+
+  if (jsonString.includes("advice_tape")) {
+    if (o.advice_tape.len === 0 || o.advice_tape.some(isNaN)) {
+      toast.error(`If you use advice_tape, 
+      it must contain at least one number, 
+      and can only contain numbers.`);
+      return false;
+    }
+  }
```
done

hackaugusto commented on playground/src/App.tsx:
```diff
@@ -16,6 +17,79 @@ async function getExample(example: string) {
+* We check the inputs and return true or false. We allow: 
+* - an empty input, and
+* - a valid JSON object containing stack_init or advice_tape (or both) 
+* iff the values are numbers
```
`iff` can also be the short hand for `if and only if` [ref](https://www.thefreedictionary.com/iff)

Dominik1999 commented:
I will rebase

hackaugusto commented on playground/src/pages/CodingEnvironment.tsx:
```diff
@@ -15,6 +16,98 @@ async function getExample(example: string) {
 
+   /**
+  * Helper function to check if input contains only numbers.
+  */ 
```
please fix the indentation. IMO we should have [prettier](https://prettier.io/docs/en/install.html) or something like it in the CI / pre-commit hooks.

Dominik1999 replied:
Indeed. I already have it, but when I let it run now, the changes in this PR grow very big. 

I suggest I will run prettier after this merge in a second PR? But happy to do otherwise

hackaugusto replied:
In that case I would just manually fix this and open the follow up later on

Dominik1999 replied:
got it

hackaugusto commented on playground/src/pages/CodingEnvironment.tsx:
```diff
@@ -15,6 +16,98 @@ async function getExample(example: string) {
+      return true;
+    }
+
+    if (!jsonString.includes("stack_init") && !jsonString.includes("advice_tape")) {
```
instead of checking to see if `stack_init` is a substring of `jsonString`. Instead it should check if `stack_init` is a key of the parsing result. (same for `advice_tape`)

Dominik1999 replied:
got it, is changed

hackaugusto reviewed (changes requested):
Please remove the `.DS_Store` file from the commit.

Also to prevent this, add it to the `.gitignore` file

hackaugusto commented on playground/src/pages/CodingEnvironment.tsx:
```diff
@@ -15,6 +16,96 @@ async function getExample(example: string) {
+    }
+    catch (e) {
+
+      const errorMessage = `Miden VM Inputs need to be 
```
This is working, but the error message is not great.

Could you please add the contents of `e`? It is a [`SyntaxError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SyntaxError), `e.message` should be enough.

Dominik1999 replied:
Isn't the errorMessage defined here the better approach? The SyntaxError is caused when the input is not a valid JSON, right?

Dominik1999 replied:
Got it, all good now.

I will add tests for all cases in #28 - which will be the next PR

hackaugusto commented on playground/src/pages/CodingEnvironment.tsx:
```diff
@@ -15,6 +16,96 @@ async function getExample(example: string) {
+
+    }
+
+    if (!Object.keys(jsonInput).includes("stack_init") && !Object.keys(jsonInput).includes("advice_tape")) {
```
humm, this doesn't seem to be caugthing the case with only `advice_tape`:

```
{
    "advice_tape": ["0"]
}
```

The error I get is:

![image](https://user-images.githubusercontent.com/310139/219375699-f1649794-2ee6-4d73-bf8b-ad0fe4c637a8.png)

Dominik1999 replied:
damn good catch, I need to create tests for all cases. Thanks. 

I need to change that in the Rust file, we want to allow Inputs with only `advice_tape`

hackaugusto reviewed (approved):
LGTM
